### ISSUE TITLE
Linux xSchedule FSEQ Playback Lag

### ISSUE THREAD

**Issue by jholmes09** [smeighan/xLights#1144]
**These are the results of stress-testing xSchedule on Lubuntu 18.04 with an internal eMMC drive (which also contains the OS) vs. a USB 2.0 stick. Logs attached & labeled.**

I attempted to play a 900-universe FSEQ with no music sync on an eMMC drive. While the log shows no errors, the playback is outputting to lights extremely slow (roughly 25% of expected speed).

I re-rendered the sequence as a musical FSEQ and the file plays almost perfectly now.
On a USB 2.0, I see consistent warnings in the log about frames taking 20-23ms to output but, since my refresh rate is 25ms (40FPS), this is fine. 

On an eMMC, I see warnings of over 25ms and visually there is some choppiness.

One interesting thing to note, I stress-tested a bunch of other musical sequences at 150 universes and 370 universes on both USB 2.0 stick and eMMC drive. The USB stick has zero warnings at both of these universe counts and no visual choppiness but the eMMC drive is taking longer than 25ms to output and shows visual choppiness. I'm surprised by this. You would think the on-board drive would be faster and benchmark testing shows eMMC as having over 4X the read-speed.

[900universe_no_audio_emmc_test_log.log](https://github.com/smeighan/xLights/files/2055952/900universe_no_audio_emmc_test_log.log)
[900universe_emmc_test_log.log](https://github.com/smeighan/xLights/files/2055958/900universe_emmc_test_log.log)
[900universe_usb_test_log.log](https://github.com/smeighan/xLights/files/2055953/900universe_usb_test_log.log)
[370universe_emmc_test_log.log](https://github.com/smeighan/xLights/files/2055956/370universe_emmc_test_log.log)
[370universe_usb_test_log.log](https://github.com/smeighan/xLights/files/2055957/370universe_usb_test_log.log)
[150universe_emmc_test_log.log](https://github.com/smeighan/xLights/files/2055954/150universe_emmc_test_log.log)
[150universe_usb_test_log.log](https://github.com/smeighan/xLights/files/2055955/150universe_usb_test_log.log)

**Comment by cjd** [comment]
Can you please give us a bit of background on your setup.
Which Linux distro/version are you running?
What desktop environment are you using?
Which linux packages/binary/build are you running for xLights (ie: appimage, ubuntu package, custom compile)

**Comment by jholmes09** [comment]
Oh I'm sorry! I completely forgot to paste that.

Lubuntu 18.04 LTS
Linux 4.15.0-22-generic x86_64

Installing xLights via Ubuntu PPA
xLights 2018.14
64-Bit

**Comment by jholmes09** [comment]
Hello, I am following up on this. Do you need any further information or testing? Let me know! I’m happy to help!

**Comment by AzGilrock** [comment]
There's only one thing I see here that could be looked at and that's where you say rendering as musical was better.  The fact is rendering as musical is exactly the same as rendering for animation.  We don't do anything different.  So what you really meant was that playback is better when there is audio being streamed which I'm assuming does something to fix the timing.  That's the only thing I could see Keith looking into.  Other than that I don't know how anyone would change code to make it work better on Linux.  We know Illuminations ran over a 1000 universes and I believe they were on Windows.  We don't have any code that says hey this is USB lets run it faster.  Your just establishing hardware limitations that we could relay to others.

**Comment by jholmes09** [comment]
Thanks for sharing your insight Gil. My auxiliary information was merely to reflect the fact that, on specs alone, eMMC should be faster than USB, but the log suggests otherwise. Conversely, there are no error logs when playing a non-musical sequence but visually I see its running at 25% speed.

To that end, let me know if you need assistance in testing more non-musical FSEQs vs. Musical FSEQs.

I suppose, regarding your other comment, it might be worth considering different ways to read the data from the drive (whatever that drive is). Has there been any discussion of preloading chunks of data into RAM so the time-sensitive part of reading/sending out ethernet is unencumbered?

**Comment by AzGilrock** [comment]
Well my comments are more of a drive-by because I probably wouldn't be the one working on this.  I've got to finish the 3D stuff I'm deep into and I've only barely touched the xSchedule code.

Are you running on a real computer or is this a small single board computer like a BeagleBone Black?  My BBB has eMMC but I've not heard about it in normal PCs.

**Comment by jholmes09** [comment]
Hi Gil, I'm running an SBC style system.

**Comment by keithsw1111** [comment]
Can you check this with .19?

**Comment by keithsw1111** [comment]
Jeff is this one fixed?

**Comment by jholmes09** [comment]
Sorry! Got distracted by the hiccuping issue. Let me re-run the test and I’ll get back to you.

**Comment by jholmes09** [comment]
@keithsw1111 this issue is still present.

I ran a 700 Universe animation sequence (no music) and found it was lagging badly. The time counter on screen was roughly half the speed of the system time.

I then ran the same sequence but disabled most of the outputs in xLights. I left only 100U outputting.
This time, the sequence ran twice as fast as it was supposed to. Looking at the time counter, it would skip time, stating 00:01, 00:02, 00:04, 00:05, 00:07, etc. etc.
